File: go_scripts/pkg/pipeline/url_rewriter.go
package pipeline

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/nyamiiko/script_hub/go_scripts/pkg/hub"
)

var mockReplacements = map[string]string{
	`https://(?:raw\.githubusercontent\.com|cdn\.jsdelivr\.net/gh)/[^"'\s]+/reject-200\.txt`:      "https://cdn.jsdelivr.net/gh/nowaytouse/script_hub@master/modules/source/mocks/reject-200.txt",
	`https://(?:raw\.githubusercontent\.com|cdn\.jsdelivr\.net/gh)/[^"'\s]+/blank\.txt`:           "https://cdn.jsdelivr.net/gh/nowaytouse/script_hub@master/modules/source/mocks/blank.txt",
	`https://(?:raw\.githubusercontent\.com|cdn\.jsdelivr\.net/gh)/[^"'\s]+/reject-dict\.json`:    "https://cdn.jsdelivr.net/gh/nowaytouse/script_hub@master/modules/source/mocks/reject-dict.json",
	`https://(?:raw\.githubusercontent\.com|cdn\.jsdelivr\.net/gh)/[^"'\s]+/reject-img\.gif`:      "https://cdn.jsdelivr.net/gh/nowaytouse/script_hub@master/modules/source/mocks/reject-img.gif",
	`https://(?:raw\.githubusercontent\.com|cdn\.jsdelivr\.net/gh)/[^"'\s]+/blank_dict\.json\.js`: "https://cdn.jsdelivr.net/gh/nowaytouse/script_hub@master/modules/source/mocks/blank_dict.json.js",
	`https://(?:raw\.githubusercontent\.com|cdn\.jsdelivr\.net/gh)/[^"'\s]+/blank\.gif`:           "https://cdn.jsdelivr.net/gh/nowaytouse/script_hub@master/modules/source/mocks/blank.gif",
	`https://(?:raw\.githubusercontent\.com|cdn\.jsdelivr\.net/gh)/[^"'\s]+/blank_dict\.json`:     "https://cdn.jsdelivr.net/gh/nowaytouse/script_hub@master/modules/source/mocks/blank_dict.json",
	`https://raw\.githubusercontent\.com/([^/]+)/([^/]+)/([^/]+)/([^"'\s]+\.[a-zA-Z0-9]+)`:                                    `https://cdn.jsdelivr.net/gh/$1/$2@$3/$4`,
	`(^|[^/])https://github\.com/([^/]+)/([^/]+)/raw/([^/]+)/([^"'\s]+\.[a-zA-Z0-9]+)`:                                        `$1https://cdn.jsdelivr.net/gh/$2/$3@$4/$5`,
	`(^|[^/])https://github\.com/([^/]+)/([^/]+)/releases/download/([^/]+)/([^"'\s]+\.[a-zA-Z0-9]+)`:                          `$1https://ghproxy.net/https://github.com/$2/$3/releases/download/$4/$5`,
}

type urlReplacement struct {
	regex *regexp.Regexp
	repl  string
}

var compiledReplacements []urlReplacement

func init() {
	for pattern, repl := range mockReplacements {
		compiledReplacements = append(compiledReplacements, urlReplacement{
			regex: regexp.MustCompile("(?i)" + pattern),
			repl:  repl,
		})
	}
}

func isGithubSource(path string) bool {
	parts := strings.Split(filepath.Clean(path), string(os.PathSeparator))
	for _, part := range parts {
		if part == "github" {
			return true
		}
	}
	return false
}

func CopyGithubVariants() {
	hub.Info("Starting CopyGithubVariants...")
	dirsToCopy := []string{
		filepath.Join(hub.MODULES_DIR, "surge", "amplify_nexus"),
		filepath.Join(hub.MODULES_DIR, "surge", "head_expanse"),
		filepath.Join(hub.MODULES_DIR, "shadowrocket", "amplify_nexus"),
		filepath.Join(hub.MODULES_DIR, "shadowrocket", "head_expanse"),
		filepath.Join(hub.RULESETS_DIR, "RULE-SET"),
		filepath.Join(hub.RULESETS_DIR, "AdBlock"),
	}

	skipCopy := map[string]bool{
		"🚫 Universal Ad-Blocking Rules Dependency Component PROMAX (Kali-style).sgmodule": true,
		"📱 Universal Ad-Blocking Rules (PROMAX Lite).sgmodule":                         true,
		"🚫 Universal Ad-Blocking Rules Dependency Component PROMAX (Kali-style).module":   true,
		"📱 Universal Ad-Blocking Rules (PROMAX Lite).module":                           true,
	}

	for _, dir := range dirsToCopy {
		hub.Info(fmt.Sprintf("Scanning dir to copy: %s", dir))
		if !hub.ValidateDirExists(dir, "") {
			hub.Warn(fmt.Sprintf("Directory does not exist: %s", dir))
			continue
		}
		githubDir := filepath.Join(dir, "github")
		hub.EnsureDir(githubDir)

		files, err := os.ReadDir(dir)
		if err != nil {
			hub.Warn(fmt.Sprintf("Failed to read dir %s: %v", dir, err))
			continue
		}

		reverseRegex := regexp.MustCompile(`(^|[^/])https://cdn\.jsdelivr\.net/gh/([^/]+)/([^/@]+)@([^/]+)/([^"'\s]+\.[a-zA-Z0-9]+)`)

		for _, f := range files {
			if f.IsDir() || f.Name() == "github" || skipCopy[f.Name()] {
				continue
			}
			ext := filepath.Ext(f.Name())
			if ext == ".sgmodule" || ext == ".module" || ext == ".list" {
				srcPath := filepath.Join(dir, f.Name())
				destPath := filepath.Join(githubDir, f.Name())
				content := hub.ReadFileString(srcPath)
				
				// Fix empty commit bouncing ("cdj前后缀" issue): 
				// The root file might have already been rewritten to CDN in a previous run.
				// Before copying to the github/ variants directory, reverse the CDN URLs back to raw.githubusercontent.com.
				content = reverseRegex.ReplaceAllString(content, "${1}https://raw.githubusercontent.com/${2}/${3}/${4}/${5}")
				
				hub.SafeWriteFile(destPath, content, true)
				hub.Success(fmt.Sprintf("Copied raw variant: %s -> %s", f.Name(), githubDir))
			}
		}
	}
}

func RunUrlRewrites(directory string) int {
	if !hub.ValidateDirExists(directory, "") {
		return 0
	}
	directory = filepath.Clean(directory)
	modifiedCount := 0

	err := filepath.Walk(directory, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() {
			if strings.Contains(path, ".git") {
				return filepath.SkipDir
			}
			if isGithubSource(path) {
				return filepath.SkipDir
			}
			return nil
		}
		ext := filepath.Ext(path)
		validExts := map[string]bool{
			".sgmodule": true, ".module": true, ".list": true,
			".conf": true, ".json": true, ".html": true, ".md": true,
		}
		if !validExts[ext] {
			return nil
		}

		content := hub.ReadFileString(path)
		origContent := content

		for _, rep := range compiledReplacements {
			content = rep.regex.ReplaceAllString(content, rep.repl)
		}

		if content != origContent {
			hub.SafeWriteFile(path, content, true)
			modifiedCount++
		}
		return nil
	})

	if err != nil {
		hub.Error(fmt.Sprintf("Error walking directory %s: %v", directory, err))
	}
	return modifiedCount
}

func RunAllUrlRewrites() int {
	CopyGithubVariants()
	count := 0
	count += RunUrlRewrites(hub.MODULES_DIR)
	count += RunUrlRewrites(hub.RULESETS_DIR)
	count += RunUrlRewrites(hub.ROOT)
	hub.Info(fmt.Sprintf("URL rewrite completed: %d files modified.", count))
	return count
}

File: go_scripts/rust_processor/src/url_rewriter.rs
use regex::Regex;
use std::collections::HashSet;
use std::fs;
use std::path::Path;
use walkdir::WalkDir;
use once_cell::sync::Lazy;

static MOCK_REPLACEMENTS: Lazy<Vec<(Regex, &'static str)>> = Lazy::new(|| {
    vec![
        (Regex::new(r#"(?i)https://(?:raw\.githubusercontent\.com|cdn\.jsdelivr\.net/gh)/[^"'\s]+/reject-200\.txt"#).unwrap(), "https://cdn.jsdelivr.net/gh/nowaytouse/script_hub@master/modules/source/mocks/reject-200.txt"),
        (Regex::new(r#"(?i)https://(?:raw\.githubusercontent\.com|cdn\.jsdelivr\.net/gh)/[^"'\s]+/blank\.txt"#).unwrap(), "https://cdn.jsdelivr.net/gh/nowaytouse/script_hub@master/modules/source/mocks/blank.txt"),
        (Regex::new(r#"(?i)https://(?:raw\.githubusercontent\.com|cdn\.jsdelivr\.net/gh)/[^"'\s]+/reject-dict\.json"#).unwrap(), "https://cdn.jsdelivr.net/gh/nowaytouse/script_hub@master/modules/source/mocks/reject-dict.json"),
        (Regex::new(r#"(?i)https://(?:raw\.githubusercontent\.com|cdn\.jsdelivr\.net/gh)/[^"'\s]+/reject-img\.gif"#).unwrap(), "https://cdn.jsdelivr.net/gh/nowaytouse/script_hub@master/modules/source/mocks/reject-img.gif"),
        (Regex::new(r#"(?i)https://(?:raw\.githubusercontent\.com|cdn\.jsdelivr\.net/gh)/[^"'\s]+/blank_dict\.json\.js"#).unwrap(), "https://cdn.jsdelivr.net/gh/nowaytouse/script_hub@master/modules/source/mocks/blank_dict.json.js"),
        (Regex::new(r#"(?i)https://(?:raw\.githubusercontent\.com|cdn\.jsdelivr\.net/gh)/[^"'\s]+/blank\.gif"#).unwrap(), "https://cdn.jsdelivr.net/gh/nowaytouse/script_hub@master/modules/source/mocks/blank.gif"),
        (Regex::new(r#"(?i)https://(?:raw\.githubusercontent\.com|cdn\.jsdelivr\.net/gh)/[^"'\s]+/blank_dict\.json"#).unwrap(), "https://cdn.jsdelivr.net/gh/nowaytouse/script_hub@master/modules/source/mocks/blank_dict.json"),
        (Regex::new(r#"(?i)https://raw\.githubusercontent\.com/([^/]+)/([^/]+)/([^/]+)/([^"'\s]+\.[a-zA-Z0-9]+)"#).unwrap(), "https://cdn.jsdelivr.net/gh/${1}/${2}@${3}/${4}"),
        (Regex::new(r#"(?i)(^|[^/])https://github\.com/([^/]+)/([^/]+)/raw/([^/]+)/([^"'\s]+\.[a-zA-Z0-9]+)"#).unwrap(), "${1}https://cdn.jsdelivr.net/gh/${2}/${3}@${4}/${5}"),
        (Regex::new(r#"(?i)(^|[^/])https://github\.com/([^/]+)/([^/]+)/releases/download/([^/]+)/([^"'\s]+\.[a-zA-Z0-9]+)"#).unwrap(), "${1}https://ghproxy.net/https://github.com/${2}/${3}/releases/download/${4}/${5}"),
    ]
});

static REVERSE_CDN_REGEX: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r#"(?i)(^|[^/])https://cdn\.jsdelivr\.net/gh/([^/]+)/([^/@]+)@([^/]+)/([^"'\s]+\.[a-zA-Z0-9]+)"#).unwrap()
});

pub fn get_semantic_content_pub(text: &str) -> String {
    let mut filtered = Vec::new();
    for line in text.lines() {
        let stripped = line.trim();
        if stripped.is_empty() {
            continue;
        }
        if stripped.starts_with('#') || stripped.starts_with("//") {
            continue;
        }
        filtered.push(stripped);
    }
    filtered.join("\n")
}

fn safe_write_file(path: &Path, content: &str) -> std::io::Result<()> {
    if path.exists() {
        if let Ok(old_content) = fs::read_to_string(path) {
            let old_stripped = get_semantic_content_pub(&old_content);
            let new_stripped = get_semantic_content_pub(content);
            if old_stripped == new_stripped {
                return Ok(());
            }
        }
    }
    // Simple write for now; atomic temp writing could be added if critical
    fs::write(path, content)
}

fn is_github_source(path: &Path) -> bool {
    for comp in path.components() {
        if let Some(s) = comp.as_os_str().to_str() {
            if s == "github" {
                return true;
            }
        }
    }
    false
}

pub fn copy_github_variants(root_dir: &str) {
    let root = Path::new(root_dir);
    let modules_dir = root.join("modules");
    let rulesets_dir = root.join("rulesets");

    let dirs_to_copy = vec![
        modules_dir.join("surge").join("amplify_nexus"),
        modules_dir.join("surge").join("head_expanse"),
        modules_dir.join("shadowrocket").join("amplify_nexus"),
        modules_dir.join("shadowrocket").join("head_expanse"),
        rulesets_dir.join("RULE-SET"),
        rulesets_dir.join("AdBlock"),
    ];

    let skip_copy: HashSet<&str> = [
        "🚫 Universal Ad-Blocking Rules Dependency Component PROMAX (Kali-style).sgmodule",
        "📱 Universal Ad-Blocking Rules (PROMAX Lite).sgmodule",
        "🚫 Universal Ad-Blocking Rules Dependency Component PROMAX (Kali-style).module",
        "📱 Universal Ad-Blocking Rules (PROMAX Lite).module",
    ].iter().cloned().collect();

    for dir in dirs_to_copy {
        if !dir.exists() {
            continue;
        }
        let github_dir = dir.join("github");
        let _ = fs::create_dir_all(&github_dir);

        if let Ok(entries) = fs::read_dir(&dir) {
            for entry in entries.flatten() {
                let file_type = entry.file_type().unwrap();
                let file_name_os = entry.file_name();
                let file_name = file_name_os.to_string_lossy();

                if file_type.is_dir() || file_name == "github" || skip_copy.contains(file_name.as_ref()) {
                    continue;
                }

                if file_name.ends_with(".sgmodule") || file_name.ends_with(".module") || file_name.ends_with(".list") {
                    let src_path = entry.path();
                    let dest_path = github_dir.join(file_name.as_ref());

                    if let Ok(content) = fs::read_to_string(&src_path) {
                        let new_content = REVERSE_CDN_REGEX.replace_all(&content, "${1}https://raw.githubusercontent.com/${2}/${3}/${4}/${5}");
                        let _ = safe_write_file(&dest_path, &new_content);
                    }
                }
            }
        }
    }
}

pub fn run_url_rewrites(directory: &str) -> i32 {
    let path = Path::new(directory);
    if !path.exists() {
        return 0;
    }

    let valid_exts: HashSet<&str> = ["sgmodule", "module", "list", "conf", "json", "html", "md"].iter().cloned().collect();
    let mut modified_count = 0;

    for entry in WalkDir::new(directory).into_iter().filter_map(|e| e.ok()) {
        let entry_path = entry.path();

        if entry.file_type().is_dir() {
            continue;
        }

        if entry_path.to_string_lossy().contains(".git") {
            continue;
        }
        if is_github_source(entry_path) {
            continue;
        }

        if let Some(ext) = entry_path.extension().and_then(|e| e.to_str()) {
            if !valid_exts.contains(ext) {
                continue;
            }
        } else {
            continue;
        }

        if let Ok(content) = fs::read_to_string(entry_path) {
            let orig_content = content.clone();
            let mut new_content = content;

            for (re, repl) in MOCK_REPLACEMENTS.iter() {
                new_content = re.replace_all(&new_content, *repl).to_string();
            }

            if new_content != orig_content {
                let _ = safe_write_file(entry_path, &new_content);
                modified_count += 1;
            }
        }
    }
    modified_count
}
